### DISCUSSION TITLE
google analytics error: @next/third-parties: GA has not been initialized

### DISCUSSION THREAD

**Discussion by mowliv** [vercel/next.js#70093] [Help] [answered]
### Summary

I've followed these steps outlined in [this doc](https://nextjs.org/docs/app/building-your-application/optimizing/third-party-libraries#google-analytics) to implement google analytics. I'm getting the runtime client console error:

`@next/third-parties: GA has not been initialized`

How do I initialize it while using this cool new library? Not outlined in the docs.

**Comment by Juneezee** [comment]
According to the source code, `GA has not been initialized` is printed when the `dataLayerName` is `undefined`. The default value of `dataLayerName` is `dataLayer` if you setup your `<GoogleAnalytics />` component correctly (https://nextjs.org/docs/app/building-your-application/optimizing/third-party-libraries#options-1).

https://github.com/vercel/next.js/blob/30ce5d4b3f585929198d12b779152f7eb4b1e089/packages/third-parties/src/google/ga.tsx#L60-L63

Did you put your `<GoogleAnalytics />` in the root layout?

**Reply by mowliv** [reply]
Thanks for your prompt reply. Yes, I did. I see dataLayerName is an optional parameter for it. I added that, but it did not help:

```
  return (
    <Html lang="en">
      <Head>
        <link
          href="https://fonts.googleapis.com/icon?family=Material+Icons"
          rel="stylesheet"
        />
        <meta httpEquiv="Content-Security-Policy" content={cspContent} />
      </Head>
      <body>
        <Main />
        <NextScript />
      </body>
      <GoogleAnalytics
        gaId={process.env.NEXT_PUBLIC_GOOGLE_ANALYTICS_ID || ''}
        dataLayerName="dataLayer"
      />
    </Html>
  );
```

When I view source, I don't see any GA content. The only 'google' text shows up in my CSP setup. And no 'next-ga'...

```
<!doctype html>
<html lang="en">
  <head>
    <style data-next-hide-fouc="true">
      body {
        display: none;
      }
    </style>
    <noscript data-next-hide-fouc="true"
      ><style>
        body {
          display: block;
        }
      </style></noscript
    >
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width" />
    <meta name="next-head-count" content="2" />
    <link
      href="https://fonts.googleapis.com/icon?family=Material+Icons"
      rel="stylesheet"
    />
    <meta
      http-equiv="Content-Security-Policy"
      content=" <<snip>> "
    />
    <link
      rel="preload"
      href="/_next/static/media/a34f9d1faa5f3315-s.p.woff2"
      as="font"
      type="font/woff2"
      crossorigin="anonymous"
      data-next-font="size-adjust"
    />
    <link rel="preload" href="/_next/static/css/pages/_app.css" as="style" />
    <link
      rel="stylesheet"
      href="/_next/static/css/pages/_app.css"
      data-n-g=""
    />
    <noscript data-n-css=""></noscript>
    <script
      defer=""
      nomodule=""
      src="/_next/static/chunks/polyfills.js"
    ></script>
    <script src="/_next/static/chunks/webpack.js" defer=""></script>
    <script src="/_next/static/chunks/main.js" defer=""></script>
    <script src="/_next/static/chunks/pages/_app.js" defer=""></script>
    <script
      src="/_next/static/chunks/pages/answers/%5BanswerId%5D.js"
      defer=""
    ></script>
    <script src="/_next/static/development/_buildManifest.js" defer=""></script>
    <script src="/_next/static/development/_ssgManifest.js" defer=""></script>
    <noscript id="__next_css__DO_NOT_USE__"></noscript>
  </head>
  <body>
     <<snip>>
  </body>
</html>
```

Console logging shows process.env.NEXT_PUBLIC_GOOGLE_ANALYTICS_ID is properly set.

**Reply by Juneezee** [reply]
I just created a new Next.js app with `npx create-next-app@latest`, then `npm install @next/third-parties@latest`. I edited my root layout and root page to the following:

`app/layout.tsx`:

```typescript
import { GoogleAnalytics } from "@next/third-parties/google";

export default function RootLayout({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) {
  return (
    <html lang="en">
      <body>
        {children}

        <GoogleAnalytics gaId="G-XYZ" />
      </body>
    </html>
  );
}
```

`app/page.tsx`:

```typescript
export default function Home() {
  return <p>Hello world</p>;
}
```

Even though I use an invalid `G-XYZ` `gaId`, I don't get any warnings or errors,

![image](https://github.com/user-attachments/assets/352f5c7c-a94e-4e7e-bceb-caa2ba670a15)

It's weird that you don't see any GA content when you view source. Do you have any other libraries or code that could be conflicting with the `<GoogleAnalytics>` component?

**Reply by mowliv** [reply]
It could be, but I'm not sure how to figure that out. Your example is with app routing, whereas mine uses pages routing. I tried moving the Google analytics tag inside the body tag, like you show -- but it didn't help. My code is in [_document.tsx](https://github.com/anandaworldwide/ananda-library-chatbot/blob/infra/google-analytics-fix/pages/_document.tsx), given the pages routing.

Other related: [_app.tsx](https://github.com/anandaworldwide/ananda-library-chatbot/blob/infra/google-analytics-fix/pages/_app.tsx), [layout.tsx](https://github.com/anandaworldwide/ananda-library-chatbot/blob/infra/google-analytics-fix/components/layout.tsx).

**Comment by mowliv** [comment]
Looks like a problem with Pages routing and the library.

I did this and picked **No** for App Router:
```
npx create-next-app@latest
npm i @next/third-parties@latest
```

Made pages_document.tsx:
```
import { Html, Head, Main, NextScript } from "next/document";
import { GoogleAnalytics } from "@next/third-parties/google";

export default function Document() {
  return (
    <Html lang="en">
      <Head />
      <body className="antialiased">
        <Main />
        <NextScript />
        <GoogleAnalytics gaId="G-XXXXXXXXXX" />
      </body>
    </Html>
  );
}
```

And nothing. I also tried with `<GA>` tag outside of `<body>`.

<img width="1519" alt="image" src="https://github.com/user-attachments/assets/f5034f28-9bde-4981-aed4-532937b474be">

**Reply by Juneezee** [reply, marked as answer]
> Made pages_document.tsx:

Can you try putting `<GoogleAnalytics>` in `_app.tsx`? That's what the docs says https://nextjs.org/docs/pages/building-your-application/optimizing/third-party-libraries#google-analytics

**Comment by mowliv** [comment]
Thanks. Made some progress by doing that. The assets now show up in the HTML source. And it sends a few events before it stops working. But in the simple test site it continuously sends events, about every six seconds. So perhaps there is another package conflicting.

**Reply by mowliv** [reply]
I switched to nextjs-google-analytics and got it completely working. Thanks for your help in trying out the package.
